### DISCUSSION TITLE
Explicit namespaces in operation names

### DISCUSSION THREAD

**Discussion by AntonLydike** [xdslproject/xdsl#463] [General]
I have started to write 

```python
from xdsl.dialects import scf, func

# ...
  scf.For(..., [
    func.Call(...)
  ]
# ...
```

instead of 

```python
from xdsl.dialects.scf import For
from xdsl.dialects.func import Call

# ...
  For(..., [
    Call(...)
  ]
# ...
```

I find that this makes it much clearer which dialect an operation comes from. It's also closer to the printed version of things.

I think we should move towards using modules to namespace operations in our code.

Especially if we get more dialects in the future, when we will have clashes of class names

**Comment by webmiche** [comment]
I agree, this looks bette.r What feels weird to me is that now, `scf` looks like the dialect, where it actually is the module. This might confused people, as we have actual dialect classes in the code.

IMO, if we really want something clean like this, we should change the dialect classes to support this functionality. Then you can import the actual dialect class and access its operations as properties.

**Reply by superlopuh** [reply]
Yeah this is good also. The way I would recommend trying this is by passing the dialect as the argument to `irdl_op_def`, so

``` python
scf = Dialect()

@irdl_op_def(dialect=scf)
class For(Operation):
  ...
```

**Reply by webmiche** [reply]
Yes that could work. I guess we could already make this work with the current way of passing operations to the dialects in lists. Note sure whether typechecking can handle it, but it feels easier to do than what you are suggesting.

Personally, I generally feel that we should make the annotators more parametric, but that's a broader point than this discussion :)

**Reply by AntonLydike** [reply]
Wow! This looks very exciting! I'd love to see this work with the type system! Although I'm not sure how this would work...
